add vm factory support

bergwolf opened the pull request:
The PR adds vm factory support to kata runtime per design in [the VM Factory plugin section](https://github.com/kata-containers/documentation/blob/master/design/kata-api-design.md#vm-factory-plugin). The vm factory controls how a new vm is created:
- direct: vm is created directly
- template: vm is created via vm template. A template vm is pre-created and saved. Later vm is just a clone of the template vm so that they readonly share a portion of initial memory (including kernel, initramfs and the kata agent). CPU and memory are hot plugged when necessary.
- cache: vm is created via vm caches. A set of cached vm are pre-created and maintained alive. New vms are created by just picking a cached vm. CPU and memory are hot plugged when necessary.

To support vm factory, virtcontainers is refactored so that:
1. hypervisor and sandbox are decoupled
2. a new VM layer is added. A "VM" is a guest without any sandbox association and can be assigned to a sandbox when necessary.
3. builtin proxy based agent can run internally without association with a sandbox.

A new kata CLI subcommand `kata-runtime factory` is added. It is used to manage vm factory init and destroy. Only direct and template factory are supported by kata CLI, because the cache factory requires a running daemon to maintain the cache vm channels. OTOH, all factories are available to use at the kata API level.

VMs created by vm factory still lack network support because right now network devices are configured **before** vm is created. We need network hotplug support (#287) to enable network configuration **after** vm is created. That said, this patch works otherwise and is large enough. So I'd like to get it merged and continue working on the network support when #287 is done.

jodh-intel commented on virtcontainers/vm.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (v *VM) assignSandbox(s *Sandbox) error {
+	// TODO: add vm symlinks
```
Incomplete?

bergwolf commented:
Removing wip and the do-not-merge tag. The PR is working and ready for review now. @sboeuf @jodh-intel @egernst @WeiZhang555 @laijs PTAL. In the meaning time, I'll fix up CI and post perf numbers.

bergwolf commented on cli/config/configuration.toml.in:
```diff
@@ -119,6 +119,18 @@ enable_iothreads = @DEFENABLEIOTHREADS@
+[factory]
+# VM templating support. Once enabled, new VMs are created from template
+# using vm cloning. They will share the same initial kernel memory by mapping
+# it readonly. When disabled, new VMs are created from scrash.
```
Good point! I'll add some text.

jodh-intel replied:
Also, we either need a new `enable_debug=` option for this section, or a comment explaining the the runtime `enable_debug` option affects factory logging?

bergwolf replied:
@jodh-intel @amshinde The purpose of the PR is to add vm factory functionality. The CLI side change is just minimal to use vm factory. I can add more CLI changes after the PR is merged.

bergwolf commented on vendor/github.com/intel/govmm/qemu/qemu.go:
```diff
@@ -1126,6 +1126,10 @@ type Memory struct {
 	// MaxMem is the maximum amount of memory that can be made available
 	// to the guest through e.g. hot pluggable memory.
 	MaxMem string
+
```
The govmm PR is not merge yet. I'll update vendor properly once that one gets merged.

bergwolf commented on virtcontainers/factory/cache/cache.go:
```diff
@@ -0,0 +1,86 @@
+
+func New(count uint, b base.FactoryBase) base.FactoryBase {
+	if count < 1 {
+		return b
```
It's actually OK to have count == 0 in which case we still return the base factory.

katabuilder commented:
PSS Measurement:
Qemu: 161919 KB
Proxy: 6752 KB
Shim: 10851 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1999732 KB

bergwolf commented:
Script to run: `echo 3 |sudo tee /proc/sys/vm/drop_caches; free -h; for I in $(seq 100); do echo {$I}th; docker run -d --runtime kata busybox top;done;free -h`

I've changed kata config to create 128MB guests by default.

w/o vm templating:
```
before test:
              total        used        free      shared  buff/cache   available
Mem:            31G        1.0G         29G        173M        722M         29G
Swap:          9.3G          4K        9.3G
```
```
after test:
              total        used        free      shared  buff/cache   available
Mem:            31G         15G         14G        190M        1.1G         15G
Swap:          9.3G          4K        9.3G
```
w/ vm templating:
```
              total        used        free      shared  buff/cache   available
Mem:            31G        1.2G         29G        253M        833M         29G
Swap:          9.3G          4K        9.3G
```
```
after test:
              total        used        free      shared  buff/cache   available
Mem:            31G        6.2G         23G        271M        1.1G         23G
Swap:          9.3G          4K        9.3G
```

So vm templating helped saving about 9GB for 100 guests, which is around 90MB~ for each 128MB guest. It's clear that vm templating can be very useful for density workload.

@egernst ^^

WeiZhang555 commented:
Wow, this looks like an amazing optimization. And it's also quite a huge patch, need some time to read~

katabuilder commented:
PSS Measurement:
Qemu: 161626 KB
Proxy: 8801 KB
Shim: 8899 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1996880 KB

grahamwhaley commented:
Memory savings look good :-)
We have a script over in CC (that at some point I'll move over into the kata repo) that took a bunch of measurements for looking at size/density: https://github.com/clearcontainers/tests/blob/master/metrics/density/footprint_data.md

I presume this measurement was taken with KSM turned off. I have a feeling that some of those savings could also be found with KSM, but I suspect:
- KSM would take longer and consume more CPU resource to find and merge the pages
- some of the shared pages in QEMU are on the host side only, where KSM is not generally enabled (the host side pages are not marked as mergeable) - so I think we will get gains here that KSM cannot currently get.

So, those memory savings look good to me :-)

grahamwhaley commented on vendor/github.com/intel/govmm/qemu/qemu.go:
```diff
@@ -1126,6 +1126,10 @@ type Memory struct {
 	// MaxMem is the maximum amount of memory that can be made available
 	// to the guest through e.g. hot pluggable memory.
 	MaxMem string
+
```
In the patch title - `hack/vendor` made me think
- that this was a patch in the https://github.com/kata-containers/runtime/tree/master/virtcontainers/hack directory
- and reminded me that we should have changed that subdir name from `hack` when we moved over VC
- and that no PRs should probably container the word `hack` ;-)

Can you drop the `hack` from the subject pls?

bergwolf replied:
That commit will be reworked once the govmm PR gets merged. The `hack` in the commit subject is to remind myself of the rework ;)

grahamwhaley commented on cli/config/configuration.toml.in:
```diff
@@ -134,6 +134,13 @@ enable_iothreads = @DEFENABLEIOTHREADS@
+# using vm cloning. They will share the same initial kernel memory by mapping
+# it readonly. When disabled, new VMs are created from scrash.
+#
+#enable_template = false
```
As this is commented out, and I presume the default is therefore false/off, you may as well make the commented out line:
```
#enable_template = true
```
as that is the most likely line to be added or uncommented I think.

bergwolf replied:
Good point!

grahamwhaley commented on cli/factory.go:
```diff
@@ -0,0 +1,94 @@
+
+var factorySubCmds = []cli.Command{
+	initFactoryCommand,
+	destroyFactoryCommand,
```
Will it be possible to have some sort of 'status' command as well - to show how many runtimes are shared, or even more important I think once we have the cache/pool, how many are cached?

bergwolf replied:
It is possible and it needs corresponding virtcontainers side of changes to support it. At this point, I would defer further improvements to future PRs since this one is already pretty big now.

bergwolf commented:
@grahamwhaley Yes, these numbers were taken when KSM is off. And yes, one advantage of vm templating over KSM is that it does not consume extra CPU resources. However, it does not share the host side qemu memory either, same as KSM. All the sharing is done at the guest memory level, mostly including kernel, initramfs, kata-agent and other process memory like systemd if that is enabled.

grahamwhaley commented on virtcontainers/qemu.go:
```diff
@@ -838,6 +850,61 @@ func (q *qemu) hotplugRemoveCPUs(amount uint32) (uint32, error) {
+	return q.hotplugAddMemory(slot, memMB)
+}
+
+func (q *qemu) hotplugAddMemory(slot uint32, amount int) error {
```
Just a thought - is this tied to the vmtemplating PR, or could it (if need be, to simplify the PRs) be pulled out into a separate PR?
Also, /cc @devimc as this feels like it might overlap or interact with some of his work?

bergwolf replied:
VM templating needs cpu and memory hotplug support so I added them together.

sboeuf replied:
I understand this is needed, but this is something that is not related to the whole refactoring, and I think you should submit this as a separate PR on top of which you should rebase this current PR.
This will help to split this huge PR and people will be more motivated in reviewing it ;)

bergwolf replied:
It is not about refactoring. It is a dependency of vm factory functionality and vm factory is in fact the only consumer of the internal hotplug APIs ATM. I thought that it would lose the whole picture for reviewers if I split it.

And for reviewing, I thought it is smaller commit in a large PR that actually helps. But it seems you are right that people would like to just review it as a whole rather than looking into each commit. And the PR is huge to be reviewed as a whole (hence I get no comments on the core idea/API/refactoring part of the PR for a week :( ). I'll split some dependency commits to another PR to see if that helps.

sboeuf replied:
@bergwolf Thanks :)
And don't worry, I'm gonna spend some time today in reviewing this PR, this is a huge piece of work and I need to be focus to be able to properly review it.
Also, don't hesitate to ping people that should review this PR. Usually if you don't ask, they will not spend the time on such big PRs :smiley:

jodh-intel replied:
Agreed. I plan to re-review this tomorrow AM too fwiw.

grahamwhaley commented on virtcontainers/factory.go:
```diff
@@ -6,6 +6,6 @@
 package virtcontainers
 
 type Factory interface {
-	GetVM(hypervisorType HypervisorType, hypervisorConfig HypervisorConfig) (*VM, error)
```
Can this whole commit be folded into the one that added factory.go in the first place? Seems this fixes something effectively 'missing' from that prior commit?

grahamwhaley commented on cli/config.go:
```diff
@@ -359,7 +359,7 @@ func newQemuHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
 }
```
Ideally I would have liked to have seen the static check (lint, cyclo) fixes in (merged into) the original commits. Not a biggie, just noting.

katabuilder commented:
PSS Measurement:
Qemu: 163764 KB
Proxy: 4632 KB
Shim: 11067 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1995640 KB

jshachm commented:
Just a brief looking into the code.(quiet a lot of code.. great  job!)

`hypervisor and sandbox are decoupled` is a really good idea ~
IMO decouple will lead more chance to remain resources .So I think clear resource after exception and proper rollback are really need.  I will cherry-pick and do a little more test on this to find out if there is any resource remaining.

Will feedback later~

sboeuf commented on virtcontainers/implementation.go:
```diff
@@ -18,16 +18,22 @@ import (
 }
 
+// SetFactory implements the VC function of the same name.
+func (impl *VCImpl) SetFactory(factory Factory) {
```
You need the mock implementation to implement this new API.

bergwolf replied:
The mock implementation has `SetFactory`.

bergwolf commented on virtcontainers/qemu.go:
```diff
@@ -62,7 +60,9 @@ type qemu struct {
 const qmpCapErrMsg = "Failed to negoatiate QMP capabilities"
 
-const defaultConsole = "console.sock"
+const qmpSocket = "qmp.sock"
```
The main point is that there is no need for two qmp sockets. If soeket file name length is a concern, I can just use `qmp` instead. But we've already shortened socket parent path and `qmp.sock` survives `BuildSocketPath` check, so I'd prefer to making the naming clear such that a socket actually ends with its `.sock` suffix.

jodh-intel commented on cli/create.go:
```diff
@@ -100,6 +101,25 @@ func create(containerID, bundlePath, console, pidFilePath string, detach bool,
+		kataLog.WithField("factory", factoryConfig).Info("load vm factory")
+		f, err := vf.NewFactory(factoryConfig, true)
+		if err != nil {
+			kataLog.WithError(err).Info("load vm factory failed")
```
I think it's worth adding a comment here explaining that if vm factory support fails, the system will "degrade" automatically to using the non-factory codepath.

In fact, I think we might want to add another variable to the config file to avoid this graceful degradation as some users only want to use VM factories:

```
[factory]

# If set, fail if VM factory support is not available.
## strict=true
```

bergwolf replied:
Yes, agreed. But again, I'd like to push all these UX changes to later PRs. This PR is mostly to add core vm factory functionalities.

jodh-intel commented on cli/factory.go:
```diff
@@ -0,0 +1,98 @@
+			f, err := vf.NewFactory(factoryConfig, true)
+			if err != nil {
+				kataLog.WithError(err).Error("load vm factory failed")
+				// ignore error
```
Why?

bergwolf replied:
It happens when we destroy a factory that does not exist in the first place.

jodh-intel commented on virtcontainers/factory/template/template.go:
```diff
@@ -0,0 +1,133 @@
+		return err
+	}
+	flags := uintptr(syscall.MS_NOSUID | syscall.MS_NODEV)
+	opts := fmt.Sprintf("size=%dM", t.config.HypervisorConfig.DefaultMemSz+8)
```
Why `+8`?

bergwolf replied:
An experience value that a vm device state file is smaller than 8MB, usually it is about 3MB~ and we need to align it by 8.

devimc commented on cli/factory.go:
```diff
@@ -0,0 +1,98 @@
+			fmt.Println("vm factory initialized")
+		} else {
+			kataLog.Error("vm factory is not enabled")
+			fmt.Println("vm factory is not enabled")
```
twice?

bergwolf replied:
kataLog does not print to stdout. But I think we should tell user if the command succeeds or not. Any suggestion how to do it properly with kataLog?

devimc commented on cli/factory.go:
```diff
@@ -0,0 +1,98 @@
+				f.CloseFactory()
+			}
+		}
+		fmt.Println("vm factory destroyed")
```
why don't use `kataLog` ?

bergwolf replied:
kataLog does not print to stdout.

amshinde commented on cli/create.go:
```diff
@@ -100,6 +101,25 @@ func create(containerID, bundlePath, console, pidFilePath string, detach bool,
 		return err
 	}
 
+	if runtimeConfig.FactoryConfig.Template {
```
@bergwolf Can we handle this in virtcontainers `CreateSandbox` itself rather than doing this in the cli?
 All these implementation details about the factory could be handled by virtcontainers itself keeping the api simple.

bergwolf replied:
kata CLI is the one reading the configuration files. And for frakti, unlike kata cli, we would only initialize the vm factory for only once, instead of doing it for every `CreateSandbox`. That is the reason factory config is not pushed into `CreateSandbox`.

amshinde commented on cli/factory.go:
```diff
@@ -0,0 +1,94 @@
+	},
+}
+
+var initFactoryCommand = cli.Command{
```
Are these commands intended to be called by `frakti`? I am trying to understand the use case here.

bergwolf replied:
No, not at all. frakti does not call any of the kata CLI commands. The CLI commands are minimal for kata cli to use vm factory and it helps testing.

amshinde commented on virtcontainers/factory/template/template.go:
```diff
@@ -103,3 +116,13 @@ func (t *template) createFromTemplateVM() (*vc.VM, error) {
 	return vc.NewVM(config)
 }
+
+func (t *template) checkTemplateVM() error {
```
This needs some comments what you are doing here and what those files mean.

bergwolf replied:
Will do.

amshinde commented on virtcontainers/factory/factory.go:
```diff
@@ -22,28 +22,39 @@ type FactoryConfig struct {
 }
 
-func NewFactory(config FactoryConfig) (vc.Factory, error) {
+func NewFactory(config FactoryConfig, fetchOnly bool) (vc.Factory, error) {
```
Please add comments for the `fetchOnly` flag here.

bergwolf replied:
Will do.

bergwolf commented on virtcontainers/factory/cache/cache.go:
```diff
@@ -0,0 +1,33 @@
+//
+// cache implements base vm factory on top of other base vm factory.
+
+package cache
```
It's in the PR description.

bergwolf commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -253,15 +250,6 @@ func (q *qemuArchBase) memoryTopology(memoryMb, hostMemoryMb uint64) govmmQemu.M
 	return memory
 }
 
-func (q *qemuArchBase) append9PVolumes(devices []govmmQemu.Device, volumes []Volume) []govmmQemu.Device {
```
Yes, agreed.

bergwolf commented:
@sboeuf The VM layer is needed because vm factory needs to work on a vm guest abstraction without sandbox association. The VM layer is on top of the hypervisor layer because hypervisor is just a set of driver interfaces that has no in-memory abstraction. We can add abstraction data structure in hypervisor.go but it would just look like the VM layer and it is not as clean as a separate VM layer.

With the VM layer, we have a proper layering that:
- A VM is a guest without sandbox association
- hypervisor.go is the hypervisor driver interface that implement a VM
- A sandbox is a collection of containers
- A VM can be assigned to a sandbox and once that is done, the sandbox becomes operational

> And about the sandbox decoupling, it's not needed since we want the complexity to be hidden behind sandbox API.

The sandbox API cannot cover vm factory. For one thing, vm factory needs to be initialized before any sandbox is created. For another thing, unlike kata CLI, frakti would just initialize the factory for only once and use it as long as the daemon is running. So it needs to be a separate API that is not coupled with the sandbox APIs.

> We can do the proper parallelization from Frakti by using CreateSandbox() with the appropriate factory, and hotplug later by creating new containers. Each container will be responsible for hotplugging the resources that we need.

I agree that it makes sense to let each container hotplug what it needs. But doing hotplug with vm factory still allows the factory to prepare VMs with different sizes. I would like to keep such ability. WDYT?

jodh-intel commented on virtcontainers/factory/base/base.go:
```diff
@@ -0,0 +1,20 @@
+	Config() vc.VMConfig
+
+	// GetBaseVM returns a paused VM created by the base factory.
+	GetBaseVM() (*vc.VM, error)
```
The main advantage of a cache is of course speed. However, we need to be able to identify when the cache should be invalidated, for example when a new rootfs image containing a security fix is released.

As such, I wonder if we need to add a timestamp into `vc.VM` that shows when the cache item was created (not when  it was used). That will give us a basic way to look for "stale images" I think?

- https://github.com/kata-containers/runtime/blob/master/cli/list.go#L69
- https://github.com/clearcontainers/runtime/wiki/Stale-containers

jodh-intel commented on virtcontainers/qemu.go:
```diff
@@ -924,8 +1012,53 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
+		}
+	}
+
+	err = q.qmpMonitorCh.qmp.ExecSetMigrateArguments(q.qmpMonitorCh.ctx, fmt.Sprintf("exec:cat>%s", q.config.DevicesStatePath))
```
Are there any issues around creating the image with "version X" of qemu and then trying to use it with "version Y" of qemu? I'm thinking about build-support as well (`qemu-lite` compared with `qemu-vanilla` compared with distro-packaged `qemu`).

bergwolf replied:
In such case, the qmp command might fail. And I would classify it as a configuration mistake... It's just like the version mismatch when you upgrade kata-runtime but do not correspondingly upgrade kata-agent and end up with certain grpc failures.

sboeuf commented:
@bergwolf 
> The sandbox API cannot cover vm factory. For one thing, vm factory needs to be initialized before any sandbox is created. For another thing, unlike kata CLI, frakti would just initialize the factory for only once and use it as long as the daemon is running. So it needs to be a separate API that is not coupled with the sandbox APIs.

I clearly understand the need that you have here, and this means you need access to the hypervisor interface so that you can run a VM and stop it to save its state, but I thought it would be better to make sure that we can do that through the hypervisor interface itself. And because you have removed dependency on the sandbox through the hypervisor interface, you should be able to do that.

> I agree that it makes sense to let each container hotplug what it needs. But doing hotplug with vm factory still allows the factory to prepare VMs with different sizes. I would like to keep such ability. WDYT?

Ok, but if you need different size of VMs, why don't you simply create them with the expected amount directly. IMO the hotplug should be reserved for the containers since they are the variable in the equation (don't know what the CRI layer will ask for in advance).

bergwolf commented:
@sboeuf 
> you need access to the hypervisor interface so that you can run a VM and stop it to save its state

Except for accessing the hypervisor interface, we also need an in-memory data structure to represent the guest without a sandbox, so that we can create it (`vm.NewVM()`) and assign it to a sandbox (`vm.assignSandbox()`). The other VM APIs (`Pause`, `Save`, `Stop`, `Resume`, `Kill`, and hotplug) are just wrappers around hypervisor interface so that we do not export the entire set of hypervisor interface.

> why don't you simply create them with the expected amount directly

For performance consideration, it is much slower to create a new VM and wait for agent to be up, than simply hotplug the requested CPU/Memory.

sboeuf commented:
@bergwolf 
>Except for accessing the hypervisor interface, we also need an in-memory data structure to represent the guest without a sandbox, so that we can create it (vm.NewVM()) and assign it to a sandbox (vm.assignSandbox()). The other VM APIs (Pause, Save, Stop, Resume, Kill, and hotplug) are just wrappers around hypervisor interface so that we do not export the entire set of hypervisor interface.

`vm.assignSandbox()` is a consequence of the way you've decoupled things. If you think the other way around, the sandbox gets the info regarding an existing hypervisor and that's all it needs. About `vm.NewVM()`, I can see that you expect to check on the agent after you started the VM, but you don't strictly need to do that. The VM could be a simple entity started through the hypervisor interface without the need to know about the agent. And it's only when you create the sandbox that you provide the VM you want for your hypervisor, which leaves the sandbox taking care of calling into the agent to check if it properly started inside the VM.

>For performance consideration, it is much slower to create a new VM and wait for agent to be up, than simply hotplug the requested CPU/Memory.

That's because you assume you have to wait for the agent to be up inside the VM when you're creating it from the factory. I don't think we need this, and as explained above, this is more something that should be checked in the context of the sandbox.

bergwolf commented:
> If you think the other way around, the sandbox gets the info regarding an existing hypervisor and that's all it needs.

Ideally it is, but we cannot create actual VM in `newSandbox()` because `s.agent.createSandbox()` must be called **after** a hypervisor pointer is available but **before** the vm is created. That is a situation vm factory cannot provide. Whatever vm factory returns, it must represent a running vm otherwise vm factory is pointless.

Even with my decoupling in the hypervisor and agent layers, sandbox and hypervisor/agent are still very tightly coupled in sandbox.go, where I have to use this `assignment` to mv an existing vm to a sandbox. And even so, the network part is still broken until we have network hotplug support.

> this is more something that should be checked in the context of the sandbox.

No, it needs to be done within the vm factory. We want to wait for the agent to be up before we stop the VM and create a snapshot of its memory. That is the point we use vm template to share memory across multiple guests -- including kernel, initramfs and agent runtime memory.

The VM layer hides all the details about hypervisor and agent implementation. It is cleaner than exposing the hypervisor and agent interfaces all together.

bergwolf commented on virtcontainers/factory.go:
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2018 HyperHQ Inc.
```
That seems to bring in circular dependency:
- `virtcontainers` needs to import `virtcontainers/factory/` for the factory interface, in api.go and sandbox.go.
- `virtcontainers/factory/api.go` needs to import `virtcontainers` for `VM` and `VMConfig`.

bergwolf commented:
@WeiZhang555 Yes, it is possible. We can let `kata-runtime create` check for existing vm factory and create new factory if there is not an existing one.

However I want to get the core part of vm factory reviewed and landed first before adding more extensions. So I have been pushing back CLI UX enhancements like this. Is it OK to you? Besides, the network part is still broken until we have network hotplug support. `kata-runtime create` cannot just use it anyway.

Ideally when vm factory code is fully working and stable, we can unify to create all VMs through it. Current VM creation w/o factory can be supported by the `direct` factory.

WeiZhang555 commented:
> However I want to get the core part of vm factory reviewed and landed first before adding more extensions. So I have been pushing back CLI UX enhancements like this. Is it OK to you? 

That makes a lot of sense, definitely OK to me. I just want to confirm the use case in future, extensions can be added later in separate PR so that it can be easier for reviewer.

sboeuf commented:
@bergwolf 
>The VM layer hides all the details about hypervisor and agent implementation. It is cleaner than exposing the hypervisor and agent interfaces all together.

That's fair!
I am glad you explained all the reasons why we need this so much, and this looks good to me.

Could you submit the memory hotplug as a separate PR as discussed earlier ? I really think this could be a good thing for people to give feedback on memory hotplug only, without having to look at this whole PR.

Actually anything that is not related to the vm factory should be submitted as separate PR if we want to help people focus on the review of the vm factory.

Thanks!

sboeuf commented:
Hi @bergwolf, just a check here, when are you planning to rebase the PR and split it ?

bergwolf commented:
@sboeuf @WeiZhang555 @jodh-intel @amshinde @devimc I've addressed existing comments and rebased the PR to master. Commits are also squashed and the last fix-CI commit is gone. PTAL.

bergwolf commented:
@sboeuf sure, I'll fix up code coverage.

bergwolf commented:
FYI, corresponding qemu change has been pushed to kata-containers/qemu/pull/2.

bergwolf commented:
Sigh, still need more UTs! I'll continue adding UTs and please do not let codecov block your reviewing!

sboeuf commented on virtcontainers/vm.go:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	err = config.Valid()
+	if err != nil {
```
Nit: could you refactor every:
```go
err = ...
if err != nil {
```
into
```go
if err := ...; err != nil {
```

sboeuf commented on virtcontainers/vm.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+// Kill stops a VM process.
+func (v *VM) Kill() error {
```
I would prefer `Stop()` than `Kill()` here since I think more about the VM semantic being start/stop. The fact the VM is a process should be a detail here, hence no reason to call it `Kill()`

sboeuf commented on virtcontainers/vm.go:
```diff
@@ -0,0 +1,233 @@
+		return err
+	}
+
+	s.hypervisor = v.hypervisor
```
I understand that we assign an existing hypervisor handler to the sandbox, but shouldn't we also assign the agent handler to this same sandbox ?

bergwolf replied:
Not, the vm factory uses a builtin proxy w/ `longliveconn` set to false. So the agent may not be reusable by a normal sandbox, which might use a normal proxy, and yes, as you can see, agent and proxy are tightly coupled in virtcontainers. Also since we are not using long live agent connections, there is no overhead with dropping the existing agent handler. In fact sandbox->agent is already usable by the time this is called.

sboeuf commented on virtcontainers/factory.go:
```diff
@@ -0,0 +1,11 @@
+
+package virtcontainers
+
+type Factory interface {
```
@bergwolf I agree with @amshinde here. The difference between `Factory` and `FactoryBase` is very thin and confusing, and unless there's a very good reason to keep it, you should simplify this with `Factory` interface only (basically what you've defined in `FactoryBase` right now).

bergwolf replied:
@sboeuf That would push the vm config comparison and CPU/memory hotplug down to each of the factory implementations. `factory/factory.go` is not just a broker but also handles vm config validation/comparison and possible CPU/memory hotplugs. It's better to do it at the factory level instead of doing the same work in each of the factory implementations.

sboeuf commented on virtcontainers/factory/cache/cache.go:
```diff
@@ -0,0 +1,83 @@
+type cache struct {
+	base base.FactoryBase
+
+	cachech   chan *vc.VM
```
s/cachech/cacheCh/

sboeuf commented on virtcontainers/factory/cache/cache.go:
```diff
@@ -0,0 +1,83 @@
+
+				select {
+				case cachech <- vm:
+				case _ = <-closed:
```
I think you can simplify with:
```go
case <-closed:
```

bergwolf commented:
@sboeuf Thanks for reviewing! I agree with all your comments except for the two I just replied. I'll fix up the rest today.

bergwolf commented:
On Tue, Jul 17, 2018 at 1:24 PM, Sebastien Boeuf
<notifications@github.com> wrote:
>
> @sboeuf commented on this pull request.
>
> ________________________________
>
> In virtcontainers/vm.go:
>
> > + // First make sure the symlinks do not exist
> + os.RemoveAll(sbSharePath)
> + os.RemoveAll(sbSockDir)
> +
> + err := os.Symlink(vmSharePath, sbSharePath)
> + if err != nil {
> + return err
> + }
> +
> + err = os.Symlink(vmSockDir, sbSockDir)
> + if err != nil {
> + os.Remove(sbSharePath)
> + return err
> + }
> +
> + s.hypervisor = v.hypervisor
>
> Yes ok this will work, but I don't like that we have agent handlers "potentially" in two different places...
>
In theory we can just use one agent, and that would require further
decoupling of agent and sandbox in sandbox.go... Right now,
sandbox/agent/hypervisor are all tightly coupled in sandbox.go.

Sorry for replying from email but I can't find a reply button for this
outdated comment anymore...

-- 
bergwolf@hyper.sh

bergwolf commented:
On Tue, Jul 17, 2018 at 1:27 PM, Sebastien Boeuf
<notifications@github.com> wrote:
> @sboeuf commented on this pull request.
>
> ________________________________
>
> In virtcontainers/factory.go:
>
>> @@ -0,0 +1,11 @@
> +// Copyright (c) 2018 HyperHQ Inc.
> +//
> +// SPDX-License-Identifier: Apache-2.0
> +//
> +
> +package virtcontainers
> +
> +type Factory interface {
>
> @bergwolf I had missed this. This makes sense and it'd be great to
> explicitly explain this from the code by commenting about it. Otherwise,
> it's not obvious what difference we have between Factory and FactoryBase
> interfaces.
>
OK, I'll document it in factory/base/base.go so that we do not confuse
users who only needs to look at virtcontainers/factory.go and do not
care about FactoryBase at all.

Cheers,
Tao

-- 
bergwolf@hyper.sh

bergwolf commented:
@sboeuf PR updated. PTAL. Hopefully I have also fixed the codecov drop...

sboeuf commented:
@amshinde @devimc @grahamwhaley @egernst it'd be nice if you could take a look ;)

bergwolf commented:
This is now blocked by kata-containers/tests/issues/513 and kata-containers/tests/issues/506. restarting CI did not help.... I know it all passed yesterday. But codecov was not happy so I had to update again.

bergwolf commented:
recheck

sboeuf commented:
@kata-containers/runtime PTAL, this is important to review this PR, so that we can move forward with it. It's been around for a long time, and @bergwolf reworked it many times. PTAL !

laijs commented:
It seems that codecov/patch is not reliable. see https://github.com/kata-containers/tests/issues/508 
And, as @sboeuf said, it's been around for a long time, and @bergwolf reworked it many times, I merge it despite the codecov/patch result.

grahamwhaley commented:
Hi @bergwolf - I know this is an old and merged item (that base VM templating/factory is merged already). I was wondering, are you planning at some point on running some metrics tests to publish maybe on vm factory gains for density and boot? I think having these numbers available for Kata might be a great thing.
I've not had time to run up the factory and try it - but the [report gen](https://github.com/kata-containers/tests/tree/master/metrics/report) might help as a first pass easy look?

Maybe we should open an Issue to track this, and then we might be able to have some collab to come up with some sort of blog etc.?
/cc @annabellebertooch @brecky777
